### DISCUSSION TITLE
IPC Addresses: Globally discoverable identifiers.

### DISCUSSION THREAD

**Discussion by adlrocha** [consensus-shipyard/consensuslab#190] [General]
## Background
In the MVP implementation of IPC we introduced the `IPCAddress` as a way to add subnet information to plain Filecoin address (see [spec](https://github.com/protocol/ConsensusLab/blob/main/specs/hierarchical_consensus.md#Hierarchical-Address) for an outdated description of this address). In the end, an `IPCAddress` is the combination of a [`SubnetID`]() and a raw [Filecoin address](https://spec.filecoin.io/appendix/address/). `IPCAddress` is used to add context information of a subnet to an address when need to disambiguate, for instance, in the `from` and `to` fields of a cross-net message.

With FVM, a new type of address is introduced to the Filecoin network to support the creation of new user-defined actor addressing schemes. The f4 address has been mentioned in [some discussions](https://github.com/filecoin-project/FIPs/discussions/419#discussioncomment-3602675) as a way to implement `IPCAddress`. 

In our efforts to move IPC into production, and as we review the design we have for IPC, it may be worth re-considering the design of `IPCAddress` (and potential `SubnetID`) to make them more general and interoperable. 

## Requirements
At this point, an `IPCAddress` is serialized as any other Filecoin address (initially we even considered giving it the f04 address space before the FVM proposal came), however we may not need to constraint ourselves to represent an `IPCAddress` as a canonical fielcoin address, and we may want to make it an identifier to universally identify resources in any IPC subnet (or any blockchain?). 

What IPC requires from an `IPCAddress` is a way to: 
- Add subnet information to any address (externally owned or on-chain). 
- A way to create an `IPCAddress` from a raw address and viceversa.
- Have a "human-readable" serialization so wallets and applications have an "amenable" way of handling these addresses.

## RFP
The goal of the discussion is to collect proposals and ideas around the design of IPC addresses, and to agree on a design. A few of the questions that we should solve with the design: 
- Do we want to follow Filecoin addresses encodings? 
- Can we leverage f04 addresses/delegated addresses for the design of `IPCAddress`.
- Should `IPCAddress` support raw addresses that are not Filecoin addresses? If yes, how are we going to determine the type of raw address wrapped into the `IPCAddress`? 
- Can we use `IPCAddress` to identify other resources that are not just addresses (like blocks, transactions, etc.)? Is there value in this?
- Are we happy with the current design of `SubnetID`? Does it cover all our needs?

**Comment by anorth** [comment]
I am excited by the potential to solve address ambiguity across all of web3, not just Filecoin. I would really like to see something universal (which implies not caring about Filecoin address encodings, except to embed them in something like a multiaddr).
